crates/valence_command/src/arg_parser.rs

```diff
@@ -0,0 +1,970 @@
+            }))
+        } else {
+            Ok(PossiblyRelative::Absolute(T::arg_from_string(string)?))
+        }
+    }
+
+    fn display() -> Parser {
+        T::display()
+    }
+}
+
+impl<T: Default> Default for PossiblyRelative<T> {
+    fn default() -> Self {
+        PossiblyRelative::Absolute(T::default())
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub struct BlockPos {
+    x: PossiblyRelative<i32>,
+    y: PossiblyRelative<i32>,
+    z: PossiblyRelative<i32>,
+}
+
+impl CommandArg for BlockPos {
+    type Result = BlockPos;
+
+    fn arg_from_string(string: String) -> Result<Self::Result, CommandArgParseError> {
+        let mut split = string.split(' ');
+        let x = PossiblyRelative::<i32>::arg_from_string(split.next().unwrap().to_string())?;
+        let y = PossiblyRelative::<i32>::arg_from_string(split.next().unwrap().to_string())?;
+        let z = PossiblyRelative::<i32>::arg_from_string(split.next().unwrap().to_string())?;
+
+        Ok(BlockPos { x, y, z })
+    }
+    fn len() -> ArgLen {
+        ArgLen::Exact(3)
+    }
+
+    fn display() -> Parser {
+        Parser::BlockPos
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub struct ColumnPos {
+    x: PossiblyRelative<i32>,
+    y: PossiblyRelative<i32>,
+    z: PossiblyRelative<i32>,
+}
+
+impl CommandArg for ColumnPos {
+    type Result = ColumnPos;
+
+    fn arg_from_string(string: String) -> Result<Self::Result, CommandArgParseError> {
+        let mut split = string.split(' ');
+        let x = PossiblyRelative::<i32>::arg_from_string(split.next().unwrap().to_string())?;
+        let y = PossiblyRelative::<i32>::arg_from_string(split.next().unwrap().to_string())?;
+        let z = PossiblyRelative::<i32>::arg_from_string(split.next().unwrap().to_string())?;
+
+        Ok(ColumnPos { x, y, z })
+    }
+
+    fn len() -> ArgLen {
+        ArgLen::Exact(3)
+    }
+
+    fn display() -> Parser {
+        Parser::ColumnPos
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Default)]
+pub struct Vec3 {
+    pub x: PossiblyRelative<f32>,
+    pub y: PossiblyRelative<f32>,
+    pub z: PossiblyRelative<f32>,
+}
+
+impl CommandArg for Vec3 {
+    type Result = Vec3;
+
+    fn arg_from_string(string: String) -> Result<Self::Result, CommandArgParseError> {
+        let mut split = string.split(' ');
+        let x = PossiblyRelative::<f32>::arg_from_string(split.next().unwrap().to_string())?;
+        let y = PossiblyRelative::<f32>::arg_from_string(split.next().unwrap().to_string())?;
+        let z = PossiblyRelative::<f32>::arg_from_string(split.next().unwrap().to_string())?;
+
+        Ok(Vec3 { x, y, z })
+    }
+
+    fn len() -> ArgLen {
+        ArgLen::Exact(3)
+    }
+
+    fn display() -> Parser {
+        Parser::Vec3
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Default)]
+pub struct Vec2 {
+    x: PossiblyRelative<f32>,
+    y: PossiblyRelative<f32>,
+}
```

Comment:
Are the vec types supposed to use `f32` instead of `f64`?

Reply:
That's the difference between a vec and a dvec